### DISCUSSION TITLE
Func<T1?,T2?> should not warn when assigned Func<int, string>

### DISCUSSION THREAD

**Discussion by HobbsCode** [dotnet/roslyn#63177] [Q&A]
**Version Used**: 
.NET 6.0.302 

**Steps to Reproduce**:

Passing a Func<int, string> to a Func<T1?, T2?> causes a warning but only when both T1 and T2 are nullable.
I don't think there should be any warning.

```
    public static void Main() 
    {        
        Func<int, string> f = i => i.ToString();
        
        Foo_1(f); 
        Foo_2(f); 
        Foo_3(f);
        Foo_4(f); // Warning!
    }
    
    public static void Foo_1<T1, T2>(Func<T1 , T2 > func){} // OK.
    public static void Foo_2<T1, T2>(Func<T1 , T2?> func){} // OK.
    public static void Foo_3<T1, T2>(Func<T1?, T2 > func){} // OK.
    public static void Foo_4<T1, T2>(Func<T1?, T2?> func){} // Both T1 and T2 nullable causes warning. 
```

[sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQAxAHYCuANhQIbAUwAEMJt9hhGAzA9gwMINCAbwaCCo0dixiJDEbImEFEjAFYAPAEsSAFwA0PHAAYAfAwBmDALwNN1s5oB0AFQgBlHVG0BzABQBKAG4lZRlQgDEICAB9HF9zILDlSJiseMSQ5KjoznTg8QjslDyGAHpShgB1aigSHwBCTIYAXyamrkMkHhQGFNj1ZxwDZywTXzUBnAZh6TNzMhIwfyFmsoqAeQBpR3buDBwujB6+rEmZsYnB6YYRgH45haWVtYYtnYLJPYPu3uzOM5uo3GGkGtxmDAei2Wq3Kr22u06Pz6KABIwuIJwYMB9wsj2hLwAQhAdAALG5TagkAAmgIY5CorEYYGoZAAzjBWQwAO41OokbyOWRtAjNIA==)

**Expected Behavior**:
No warnings.

**Actual Behavior**:
Warning.

**Comment by CyrusNajmabadi** [comment]
this is because your argument type is int.  so the inferred type for `T1` is `int` which as also the case for `T1?`  (since this is an unconstrained type parameter).  If you had `Func<string, string>` you would see additional warnings as `string` vs `string?` are different, including for unconstrained type parameters.

**Reply by svick** [reply]
Then why does `Foo_2` not warn? It's also passing an argument of type `Func<int, string>` as a parameter of type `Func<int, string?>`.

**Reply by CyrusNajmabadi** [reply]
because that's totally safe.  your function says it always returns a string.  Foo_2 says i need a function that can return a `string?`  a `string` is always a safe thing to return to something that expects a `string?`.

**Comment by RikkiGibson** [comment]
The behavior seems off to me here. If we do something like `Func<int, string?> f2 = f;`, which is converting to the same type as the parameter of `Foo_4<int, string>`, we don't get a warning. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAGIYBXFFCYFAUwAIaYrbDCMBmO7OgYTsIDedfgWHDsWEWLpDpYwnLEYArAB5YCOFxwAGAHx0AZnQC8dKKYNQAdABUA9gGUEAJ1gBzABQBKANwLFKUCVdRhNbR0AfgNDSTNDf1FAgDF7ewB9ABZPQz86AHp8ugB1CBcYDwBCAMVUjOzYvMK6ewBrGroAXw6Ojm0ANi5Mujqs1VscLVssPU8QiciprGijUhgAY28BToKigCF7BAALOgm6CBgAE1PJMgpmenWIUgBnGhe6AHcyiph3a2kPQInSAA===).
